domain/cloud/state/state.go

```diff
@@ -73,6 +75,234 @@ func (st *State) ListClouds(ctx context.Context, filter *Filter) ([]cloud.Cloud,
+FROM cloud_region_defaults
+INNER JOIN cloud_region
+ON cloud_region.uuid = cloud_region_defaults.region_uuid
+INNER JOIN cloud
+ON cloud_region.cloud_uuid = cloud.uuid
+WHERE cloud.name = ?
+`
+
+	return defaults, db.StdTxn(ctx, func(ctx context.Context, tx *sql.Tx) error {
+		rows, err := tx.QueryContext(ctx, stmt, cloudName)
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		var regionName, key, value string
+		for rows.Next() {
+			if err := rows.Scan(&regionName, &key, &value); err != nil {
+				return errors.Trace(stderrors.Join(err, rows.Close()))
+			}
+			store, has := defaults[regionName]
+			if !has {
+				store = map[string]string{}
+				defaults[regionName] = store
+			}
+			store[key] = value
+		}
+		return nil
+	})
+}
+
+// UpdateCloudRegionDefaults is responsible for updating default config values
+// for a cloud region. This function will allow the addition and updating of
+// attributes. Attributes can also be removed by keys if they exist for the
+// current cloud. If the cloud or region is not found an error that satisfied
+// NotValid is returned.
+func (st *State) UpdateCloudRegionDefaults(
+	ctx context.Context,
+	cloudName string,
+	regionName string,
+	updateAttrs map[string]string,
+	removeAttrs []string,
+) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	deleteBinds, deleteVals := database.SliceToPlaceholder(removeAttrs)
+	deleteStmt := fmt.Sprintf(`
+DELETE FROM cloud_region_defaults
+WHERE key IN (%s)
+AND region_uuid = (SELECT cloud_region.uuid
+                   FROM cloud_region
+                   INNER JOIN cloud
+                   ON cloud_region.cloud_uuid = cloud.uuid
+                   WHERE cloud.name = ?
+                   AND cloud_region.name = ?)
+`[1:], deleteBinds)
+
+	upsertBinds, upsertVals := database.MapToMultiPlaceholder(updateAttrs)
+	upsertStmt := fmt.Sprintf(`
+CREATE TEMP TABLE cloud_region_default_dump (
+	key TEXT,
+    value TEXT
+);
+    
+INSERT INTO temp.cloud_region_default_dump VALUES %s;
+
+INSERT INTO cloud_region_defaults(region_uuid, key, value)
+SELECT (SELECT cloud_region.uuid
+        FROM cloud_region
+        INNER JOIN cloud
+        ON cloud.uuid = cloud_region.cloud_uuid
+        WHERE cloud.name = ?
+        AND cloud_region.name = ?) AS region_uuid,
+       key,
+       value
+FROM temp.cloud_region_default_dump
+-- The WHERE clause here is needed to help sqlite parser with ambiguity.
+WHERE true
+ON CONFLICT(region_uuid, key) DO UPDATE
+SET value = excluded.value
+WHERE region_uuid = excluded.region_uuid
+AND key = excluded.key;
+
+DROP TABLE temp.cloud_region_default_dump;
+`, upsertBinds)
+
+	return db.StdTxn(ctx, func(ctx context.Context, tx *sql.Tx) error {
+		if len(deleteVals) != 0 {
+			_, err := tx.ExecContext(ctx, deleteStmt, append(deleteVals, cloudName, regionName)...)
+			if err != nil {
+				return fmt.Errorf("removing cloud %q region %q default keys: %w", cloudName, regionName, err)
```

Comment:
let's stick to using juju errors Errorf() so we capture location and stacktrace etc in the error

Reply:
We should avoid using the Juju errors library Errorf as it will produce a cause error and won't support multiple %w. We are moving to a model of annotation with the library using the standard errors and fmt package.

It's a little more painful under some circumstances but gives us the best change of making clean errors that keep working in any additions Go does in the future.

Reply:
I've updated with a trace.

Reply:
Actually i've updated to annotate all the errors as they come out. This is the better approach in my opinion.